From 29239fa1b38f55b9a263b9582a43bae18cb5980d Mon Sep 17 00:00:00 2001
From: Marius Gerbershagen <marius.gerbershagen@gmail.com>
Date: Wed, 6 May 2020 20:58:20 +0200
Subject: [PATCH 1/3] unicode: fix ecl_string_case for non-ascii characters

Problem reported and fix provided by Vladimir Sedach on the ecl-devel
mailing list.
---
 src/c/character.d | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/src/c/character.d b/src/c/character.d
index a69b6e4b..c3ef328a 100644
--- a/src/c/character.d
+++ b/src/c/character.d
@@ -99,6 +99,7 @@ cl_both_case_p(cl_object c)
 int
 ecl_string_case(cl_object s)
 {
+  /* Returns 1 if string is all uppercase, -1 if all lowercase, and 0 if mixed case */
   int upcase;
   cl_index i;
   ecl_base_char *text;
@@ -106,16 +107,16 @@ ecl_string_case(cl_object s)
   switch (ecl_t_of(s)) {
 #ifdef ECL_UNICODE
   case t_string:
-    s = si_coerce_to_base_string(s);
 #endif
   case t_base_string:
-    text = (ecl_base_char*)s->base_string.self;
     for (i = 0, upcase = 0; i < s->base_string.dim; i++) {
-      if (ecl_upper_case_p(text[i])) {
+      ecl_character c = ecl_char(s, i);
+
+      if (ecl_upper_case_p(c)) {
         if (upcase < 0)
           return 0;
         upcase = +1;
-      } else if (ecl_lower_case_p(text[i])) {
+      } else if (ecl_lower_case_p(c)) {
         if (upcase > 0)
           return 0;
         upcase = -1;
-- 
2.26.2

